### PR TITLE
[MosaicGPU] fp4 sparsity support for Rubin

### PR DESCRIPTION
**Pull request by olupton**
This makes this Mosaic GPU test pass on sm_107a; the instruction descriptor must be set differently on sm_107a to sm_100a and sm_103a (https://docs.nvidia.com/cuda/developer-preview/13.4/parallel-thread-execution/index.html#tcgen05-instuction-desc-kind-mxf4-mxf4nvf4).

### CODE REVIEW

File: jax/experimental/mosaic/gpu/core.py [hunk 1]
```diff
@@ -784,7 +784,7 @@ def _launch(
     gpu.terminator()
 
 
-def _infer_arch() -> tuple[int, int]:
+def infer_arch() -> tuple[int, int]:
```
Comment (apaszke) [inline jax/experimental/mosaic/gpu/core.py]: Can you please undo the rename? Just removing the underscore does not make it a public API (core.py is not something users should touch). Just keep the old name

File: tests/mosaic/gpu_test.py [hunk 2]
```diff
@@ -3357,7 +3358,11 @@ def kernel(ctx, lhs, rhs, lhs_sparse_gmem, lhs_scales_gmem, rhs_scales_gmem, out
     y_shape = (n, k)
     y = self.prng.uniform(-1, 1, y_shape).astype(in_jax_dtype)
     out_shape = jax.ShapeDtypeStruct((m, n), out_jax_dtype)
-    meta_k = k // 4
+    arch = utils.Arch(*core.infer_arch())
```
Comment (apaszke) [inline tests/mosaic/gpu_test.py]: What's the problem here? Why do you need to repack the arch? Wouldn't sparse_group_elems query it for you anyway?

File: jax/experimental/mosaic/gpu/tcgen05.py [hunk 3]
```diff
@@ -112,8 +112,32 @@ def get_input_encoding(ty):
...
+  of a lowering context have to pass `arch` explicitly.
+  """
+  if arch is None:
+    arch = utils.get_arch()
+  assert arch.major in {10, 11}, arch
+  return int(arch.major == 10 and arch.minor >= 7)
+
+
+def sparse_group_elems(
+    element_type: ir.Type, arch: utils.Arch | None = None
```
Comment (apaszke) [inline jax/experimental/mosaic/gpu/tcgen05.py]: I think we can delete the arch args from both of those functions and have _mxf4_sparsity_version just call it

File: jax/experimental/mosaic/gpu/tcgen05.py [hunk 4]
```diff
@@ -123,17 +147,28 @@ def _create_scaled_instr_descriptor(
     transpose_a: bool,
     transpose_b: bool,
     scale_type: ir.Type,
+    mxf4_or_mxf4nvf4: bool,
```
Comment (apaszke) [inline jax/experimental/mosaic/gpu/tcgen05.py]: Can't you query it from the dtypes? Just see if they are 4 bit?

File: jax/experimental/mosaic/gpu/tcgen05.py [hunk 5]
```diff
@@ -123,17 +147,28 @@ def _create_scaled_instr_descriptor(
...
+    instr_k: int,
```
Comment (apaszke) [inline jax/experimental/mosaic/gpu/tcgen05.py]: How is instr_k specified? I don't see it listed on the call sites explicitly

File: jax/experimental/mosaic/gpu/tcgen05.py [hunk 6]
```diff
@@ -123,17 +147,28 @@ def _create_scaled_instr_descriptor(
...
-  # Bit 3 is reserved
+  # Bit 3 is the K dimension upper bit for .kind::mxf4 and .kind::mxf4nvf4
+  desc |= (encoded_k >> 1) << 3
   assert 0 <= b_scale_idx < 4
   desc |= b_scale_idx << 4  # B scale factor data ID, bits 4-5
   # Bit 6 is reserved
   desc |= get_input_encoding(a_type) << 7  # A dtype, bits 7-9
-  desc |= get_input_encoding(b_type) << 10  # B dtype, bits 10-12
+  encoded_b_type = get_input_encoding(b_type)
+  assert (encoded_b_type >> 3 - mxf4_or_mxf4nvf4) == 0
```
Comment (apaszke) [inline jax/experimental/mosaic/gpu/tcgen05.py]: I'd keep the code as before `desc |= get_input_encoding(...)` and then below do
```
assert not is_4_bit or (desc >> 11) == 0
```

File: jax/experimental/mosaic/gpu/tcgen05.py [hunk 7]
```diff
@@ -152,7 +187,9 @@ def _create_scaled_instr_descriptor(
     raise ValueError(f"M must be a multiple of 16 and <= 256, got: {m}")
   desc |= (m >> 7) << 27  # M >> 7, bits 27-28
   desc |= a_scale_idx << 29  # A scale factor data ID, bits 29-30
-  # Bit 31 is reserved
+  # Bit 31 is the lowest bit of the encoded K dimension for .kind::mxf4,
+  # .kind::mxf4nvf4, and .kind::mxf8f6f4
+  desc |= (encoded_k & 1) << 31
```
Comment (apaszke) [inline jax/experimental/mosaic/gpu/tcgen05.py]: Is the K encoding necessary for this change? The PR message only talks about sparsity support. For as long as we don't support any new instr_k that were not present on Blackwell, can we just drop this interpolation of  bits and split it into another change?

File: jax/experimental/mosaic/gpu/tcgen05.py [hunk 8]
```diff
@@ -765,15 +814,15 @@ def _get_offset(idx: int, idx_tiling: tuple[int, ...], strides: tuple[int, ...])
...
-      meta_cols_per_instr = instr_k // sparse_group_elems * 2 // 16
-      instrs_per_col_pair = 2 // meta_cols_per_instr
-      sp_selector = k_step % instrs_per_col_pair
+      # Sparse metadata is organised as 32-row columns of 32-bit cells in TMEM; 4 bits
+      # of metadata map to 2 (1:2 sparsity), 4 (2:4 sparsity) or 8 (4:8 sparsity)
+      # of the `instr_k` columns in A consumed by each tcgen05.mma.sp instruction.
+      meta_cols_per_instr = instr_k // sparse_group_elems(a_element_type) * 2 // 16
+      # The address is a column pair and the selector picks a half of it.
+      meta_col = k_step * meta_cols_per_instr
+      sp_selector = meta_col % 2
```
Comment (apaszke) [inline jax/experimental/mosaic/gpu/tcgen05.py:809]: Are you saying that the old code was unnecessary and instrs_per_col_pair always evaluated to 2?

File: jax/experimental/mosaic/gpu/tcgen05.py [hunk 9]
```diff
@@ -164,7 +201,11 @@ def get_input_encoding(ty):
       return 1
     else:
       raise NotImplementedError(f"Unsupported input dtype: {ty}")
-  return _create_scaled_instr_descriptor(get_input_encoding, *args, **kwargs)
+  def get_k_encoding(k: int, sparse: bool) -> int:
+    if k != 32 * (1 + sparse):
+      raise NotImplementedError(f"k: {k} {'sparse' if sparse else 'dense'} support not implemented")
+    return 0
+  return _create_scaled_instr_descriptor(get_input_encoding, get_k_encoding, *args, mxf4_or_mxf4nvf4=False,  **kwargs)
```
Comment (apaszke) [inline jax/experimental/mosaic/gpu/tcgen05.py]: Does this really pass in instr_k? where?

File: jax/experimental/mosaic/gpu/tcgen05.py [hunk 10]
```diff
@@ -782,7 +831,7 @@ def _get_offset(idx: int, idx_tiling: tuple[int, ...], strides: tuple[int, ...])
       assert sp_selector in {None, 0}  # Scaled instr descriptor has no selector
       i_desc = create_scaled_instr_descriptor(
           m * num_cta, n * num_cta, a_element_type, b_element_type,
-          scale_id, scale_id, a_transpose, b_transpose
+          scale_id, scale_id, a_transpose, b_transpose, instr_k=instr_k,
```
Comment (olupton) [inline jax/experimental/mosaic/gpu/tcgen05.py]: This is where `instr_k` was passed in (now removed).

File: jax/experimental/mosaic/gpu/core.py (hunk 1 above)
Comment (olupton) [reply jax/experimental/mosaic/gpu/core.py]: Sure, will do.

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 5 above)
Comment (olupton) [reply jax/experimental/mosaic/gpu/tcgen05.py]: I commented at the call site: https://github.com/jax-ml/jax/pull/39927/changes#r3783825920 (but it is now removed)

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 9 above)
Comment (olupton) [reply jax/experimental/mosaic/gpu/tcgen05.py]: Commented at the call site: https://github.com/jax-ml/jax/pull/39927/changes#r3783825920 (but it is now removed)

File: tests/mosaic/gpu_test.py (hunk 2 above)
Comment (olupton) [reply tests/mosaic/gpu_test.py]: The trouble is that it only queries the arch if there's an active lowering context, and there isn't one here.

It seemed like having a silent fallback to querying the current JAX device if there isn't an active context might be surprising. But if you prefer that path, I can change it.

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 3 above)
Comment (olupton) [reply jax/experimental/mosaic/gpu/tcgen05.py]: Commented here about this: https://github.com/jax-ml/jax/pull/39927/changes#r3784234074

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 4 above)
Comment (olupton) [reply jax/experimental/mosaic/gpu/tcgen05.py]: That would work for the current set of plumbed types, but in general I believe it doesn't work because this flag (and the descriptor encoding) matches the `.kind` on the instruction, and E2M1 is allowed in both: https://docs.nvidia.com/cuda/developer-preview/13.4/parallel-thread-execution/index.html#tcgen05-instuction-desc-kind-mxf8f6f4

I can change it if you like (and assert no 4-bit types going through the `mxf8f6f4` path), it just seemed logical to set this flag in the `kind`-specific code path.

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 8 above)
Comment (olupton) [reply jax/experimental/mosaic/gpu/tcgen05.py:809]: The old code didn't work for `sparse_group_elems=4` and `instr_k=128` because you got `meta_cols_per_instr=4` and, therefore, `instrs_per_col_pair=0` and a `ZeroDivisionError`. So something had to change.

I don't think `instrs_per_col_pair` always evaluated to 2: for example for fp8 we had `instr_k=64` and `sparse_group_elems=4` giving `meta_cols_per_instr=2` and `instrs_per_col_pair=1` (and `sp_selector=0` as required -- https://docs.nvidia.com/cuda/developer-preview/13.4/parallel-thread-execution/index.html#tcgen05-sparse-matrices-sparsity-selector-kind-f8f6f4-mxf8f6f4-m64). Blackwell fp4 followed the same path.

For bf16 and `instr_k=32` we get `meta_cols_per_instr` of 1 and a `sp_selector` that alternates between 0 and 1 in both versions of the code.

I found the new one clearer because it separates the logical/straightforward `meta_col` calculation from the lower-level detail that you have to pass a pair of (`base_addr + meta_col - sp_selector`, `sp_selector`) rather than just `base_addr + meta_col` when dealing with odd values of `meta_col`.

Do you see a problem?

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 6 above)
Comment (olupton) [reply jax/experimental/mosaic/gpu/tcgen05.py]: Sure, will change.

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 7 above)
Comment (olupton) [reply jax/experimental/mosaic/gpu/tcgen05.py]: Will do. Originally I was thinking of plumbing larger K through but I didn't finish it and thought leaving the encoding in might give you some more context.

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 4 above)
Comment (apaszke) [reply jax/experimental/mosaic/gpu/tcgen05.py]: We have no plans to support anything but the f8 (and maybe f6) in the mxf8f6f4 path. I don't see how it's useful for f4 given that you can just use mxf4 and have less padding?

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 8 above)
Comment (apaszke) [reply jax/experimental/mosaic/gpu/tcgen05.py:809]: I'm just confused that we're taking %2 here, since 1 is not always a valid sp_selector value? But I guess you're saying that in those cases meta_col will always be a multiple of 2?

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 3 above)
Comment (apaszke) [reply jax/experimental/mosaic/gpu/tcgen05.py]: Hmm this link doesn't take me to any useful comment?

Comment (olupton) [reply jax/experimental/mosaic/gpu/tcgen05.py]: Maybe https://github.com/jax-ml/jax/pull/39927#discussion_r3784234074 ?
pasting here in case:
> The trouble is that it only queries the arch if there's an active lowering context, and there isn't one here.
> It seemed like having a silent fallback to querying the current JAX device if there isn't an active context might be surprising. But if you prefer that path, I can change it.

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 8 above)
Comment (olupton) [reply jax/experimental/mosaic/gpu/tcgen05.py:809]: Yes; I believe the cases where `sp_selector` has to be 0 _are_ the cases where you need 2 or more columns of metadata per instruction and `meta_col` will always be even.

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 4 above)
Comment (olupton) [reply jax/experimental/mosaic/gpu/tcgen05.py]: OK, I changed it.

I think the most (theoretically) interesting reason to have f4 via f8f6f4 is probably if you want to mix f4 with f6 or f8; I don't see any good reason to use it for pure f4. But if there's no practical interest at the moment, querying the A/B bitwidth should be fine.

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 3 above)
Comment (apaszke) [reply jax/experimental/mosaic/gpu/tcgen05.py]: I don't understand the reply though? You only use this argument once, in a test, where you infer it anyway. That's not a good argument to keep it? We don't expect people to pass in arch explicitly at any point?

File: jax/experimental/mosaic/gpu/tcgen05.py [hunk 11]
```diff
@@ -765,15 +802,15 @@ def _get_offset(idx: int, idx_tiling: tuple[int, ...], strides: tuple[int, ...])
...
+      # of the `instr_k` columns in A consumed by each tcgen05.mma.sp instruction.
+      meta_cols_per_instr = instr_k // sparse_group_elems(a_element_type) * 2 // 16
+      # The address is a column pair and the selector picks a half of it.
+      meta_col = k_step * meta_cols_per_instr
+      sp_selector = meta_col % 2
       sparse_addr = (
-          arith.addi(
-              a_sparse_addr, utils.c(k_step // instrs_per_col_pair * 2, i32)
-          ),
+          arith.addi(a_sparse_addr, utils.c(meta_col - sp_selector, i32)),
```
Comment (apaszke) [inline jax/experimental/mosaic/gpu/tcgen05.py]: I'm confused by this. I thought the address should be independent of the selector?

Comment (apaszke) [reply jax/experimental/mosaic/gpu/tcgen05.py]: Or I guess this is an awkward way to implement a rouding down if needed? Can you leave a comment please?

**Review by apaszke — APPROVED** [review]

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 3 above)
Comment (olupton) [reply jax/experimental/mosaic/gpu/tcgen05.py]: I pushed an update that makes it implicit.

File: jax/experimental/mosaic/gpu/tcgen05.py (hunk 11 above)
Comment (olupton) [reply jax/experimental/mosaic/gpu/tcgen05.py]: Updated the comment and made it a more explicit `// 2 * 2` for round-down-to-even.

File: jax/experimental/mosaic/gpu/core.py [hunk 12]
```diff
@@ -815,28 +814,6 @@ def _launch(
     gpu.terminator()
 
 
-def _infer_arch() -> tuple[int, int]:
```
Comment (olupton) [inline jax/experimental/mosaic/gpu/core.py:818]: This had to move to avoid circular import dependencies.

**Review by apaszke — APPROVED** [review]
